src/coord/src/catalog.rs

```diff
@@ -2502,10 +2548,52 @@ enum SerializedCatalogItem {
         create_sql: String,
         // The name "eval_env" is historical.
         eval_env: Option<SerializedPlanContext>,
-        persist_name: Option<String>,
+        // Previous versions used "persist_name" as the field name here.
+        #[serde(alias = "persist_name")]
+        table_persist_name: Option<String>,
+        source_persist_details: Option<SerializedSourcePersistDetails>,
     },
 }
 
+/// Serialized source persistence details. See `SourcePersistDesc` for an explanation of the
+/// fields.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SerializedSourcePersistDetails {
+    /// Name of the primary persisted stream of this source. This is what a consumer of the
+    /// persisted data would be interested in while the secondary stream(s) of the source are an
+    /// internal implementation detail.
+    pub primary_stream: String,
+
+    /// Persisted stream of timestamp bindings.
+    pub timestamp_bindings_stream: String,
+
+    /// Any additional details that we need to make the envelope logic stateful.
+    pub envelope_details: SerializedEnvelopePersistDetails,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum SerializedEnvelopePersistDetails {
+    Upsert,
```

Comment:
since we're going to start writing this down, we should decide now if we want a 1:1 correspondence between envelopes and entries here vs a "no extra data" (None?) variant that Upsert/AppendOnly/etc all get mapped to. I leave the actual decision to your judgement, just want to make sure we're explicitly making it a decision

Reply:
I would put an entry per envelope. Having something like `None` is tricky because we have the `NONE` envelope. Although both `NONE` and `UPSERT` envelopes will need no extra streams.

I'll add a description here in the code.